include/envoy/http/header_map.h

```diff
@@ -373,6 +378,13 @@ class HeaderMap {
   virtual void iterate(ConstIterateCb cb, void* context) const PURE;
 
   /**
+   * Iterate over a constant header map in reverse order.
+   * @param cb supplies the iteration callback.
+   * @param context supplies the context that will be passed to the callback.
+   */
+  virtual void iterateReverse(ConstIterateCb cb, void* context) const PURE;
```

Comment:
Beyond the scope of this change, but why don't we support normal STL iterators here rather than the higher-order function approach?

Reply:
@mattklein123

Reply:
IIRC I tried to figure out how to do that on top of a virtual interface and got confused and gave up.

Reply:
Based on my very recent attempt to implement this, it would look something like this:
1. define an abstract class `HeaderMap::IteratorImpl`
1. define a concrete class `HeaderMap::Iterator` (or `iterator` for consistency with the stl) that delegates to a `HeaderMap::IteratorImpl*`
1. create `HeaderMapImpl::IteratorImpl` that subclasses `HeaderMap::IteratorImpl` and delegates to a `std::list<HeaderEntryImpl>::iterator`
1. define and implement the necessary virtual `begin()`, `end()`, `rbegin()`, `rend()`, methods on `HeaderMap` and `HeaderMapImpl`

The result is that each increment and equality comparison is a virtual function call, so 2 calls per loop iteration. The current callback method only requires 1 indirect call per iteration, so I'm kind of in favor of keeping it.

Reply:
Thanks for scoping it out. I think we can stick with the existing interface for this PR, but maybe you want to document this in an issue titled "Consider moving HeaderMap to STL iterators"? That way we can debate the merits of STL iterators and have some record for posterity.